[flang][OpenMP] Document `do concurent` mapping status and next steps (#100)
ROCm/llvm-project: flang/docs/DoConcurrentConversionToOpenMP.md

#### Multi-range and perfectly-nested loops

Currently, on the `FIR` dialect level, the following 2 loops are modelled in exactly the same way:

```fortran
do concurrent(i=1:n, j=1:m)
  a(i,j) = i * j
end do
```

```fortran
do concurrent(i=1:n)
  do concurrent(j=1:m)
    a(i,j) = i * j
  end do
end do
```

Both of the above loops are modeleld as:

```mlir
fir.do_loop %arg0 = %11 to %12 step %c1 unordered {
  ...
  fir.do_loop %arg1 = %14 to %15 step %c1_1 unordered {
    ...
  }
}
```

Consequently, from the `DoConcurrentConversionPass`' perspective, both loops are treated in the same manner. Under the hood, the pass detects perfectly-nested loop nests and maps such nests as if they were multi-range loops.

#### Non-perfectly-nested loops

One limitation that the pass currently have is that it treats any intervening code in a loop nest as being disruptive to detecting that nest as a single unit. For example, given the following input:

```fortran
do concurrent(i=1:n)
  x = 41
  do concurrent(j=1:m)
    a(i,j) = i * j
  end do
end do
```

> Since there at least one statement between the 2 loop header (i.e. `x = 41`), the pass does not detect the `i` and `j` loops as a nest. Rather, the pass in that case only maps the `i` loop to OpenMP and leaves the `j` loop in its origianl form. In theory, in this example, we can sink the intervening code into the `j` loop and detect the complete nest. However, such transformation is still to be implemented in the future.

Author: > Doesn't the pass currently detect the inner loop in that situation as a separate instance of do concurrent to transform?
> ....
> I think it's for the best to not try doing that, especially for the device, but it seems to follow from the previous logic.

That indeed does not happen now. The nested loop in this case is marked to be skipped from mapping. This is why I wrote:
> Rather, the pass in that case only maps the `i` loop to OpenMP and leaves the `j` loop in its origianl form.

Let me know if this is not clear.

The above also has the consequence that the `j` variable will **not** privatized in the OpenMP parallel/target region. In other words, it will be treated as if it was a `shared` variable. For more details about privatization, see the data environment section below.

### Data environment

By default, variables that are used inside a `do concurernt` loop nest are either treated as `shared` in case of mapping to `host`, or mapped into the `target` region using a `map` clause in case of mapping to `device`. The only exception to this is the loop's iteration variable(s) of **perfect** loop nest. In that case, for each IV, we allocate a local copy as shown the by the mapping examples above.

#### Non-perfectly-nested loops' IVs

> For non-perfectly-nested loops, the IV(s) are still treated as `shared` or `map` entries as pointed out above. This **might not** be consistent with what the Fortran specficiation tells us. In particular, taking the following snippets from the spec (version 2023) into account:

Reviewer: ```suggestion
For non-perfectly-nested loops, the IVs are still treated as `shared` or `map` entries as pointed out above. This **might not** be consistent with what the Fortran specficiation tells us. In particular, taking the following snippets from the spec (version 2023) into account:
```

> § 3.35
> ------
> construct entity
> entity whose identifier has the scope of a construct

> § 19.4
> ------
>  A variable that appears as an index-name in a FORALL or DO CONCURRENT construct, or ... is a construct entity. A variable that has LOCAL or LOCAL_INIT locality in a DO CONCURRENT construct is a construct entity.
> ...
> The name of a variable that appears as an index-name in a DO CONCURRENT construct, FORALL statement, or FORALL construct has a scope of the statement or construct. A variable that has LOCAL or LOCAL_INIT locality in a DO CONCURRENT construct has the scope of that construct.

> From the above quotes, it seems there is an equivalence between the IV of a `do concurrent` loop and a variable with a `LOCAL` locality specifier (equivalent to OpenMP's `private` clause). Which means that we should probably localize/privatize a `do concurernt` loop's IV even if it is not perfectly nested in the nest we are parallelizing. For now, however, we **do not** that as pointed out previously. In the near future, we propose a middle-ground solution (see the Next steps section for more details).

Reviewer: ```suggestion
From the above quotes, it seems there is an equivalence between the IV of a `do concurrent` loop and a variable with a `LOCAL` locality specifier (equivalent to OpenMP's `private` clause). Which means that we should probably localize/privatize a `do concurernt` loop's IV even if it is not perfectly nested in the nest we are parallelizing. For now, however, we **do not** do that as pointed out previously. In the near future, we propose a middle-ground solution (see the "Next steps" section for more details).
```

## Next steps

### Delayed privatization

> So far, we emit the privatization logic for IVs inline in the paralle/target region. This is enough for out purposes right now since we don't localize/privatize any sophisticated types of variables yet. Once we have need for more advanced localization thought `do concurrent`'s locality specified (see below), delayed privatization will enable us to have a much cleaner IR. Once delayed privatization's implementation upstream is supported for the required constructs by the pass, we will move to it rather than inlined/early privatization.

Reviewer: ```suggestion
So far, we emit the privatization logic for IVs inline in the parallel/target region. This is enough for our purposes right now since we don't localize/privatize any sophisticated types of variables yet. Once we have need for more advanced localization through `do concurrent`'s locality specifiers (see below), delayed privatization will enable us to have a much cleaner IR. Once delayed privatization's implementation upstream is supported for the required constructs by the pass, we will move to it rather than inlined/early privatization.
```

### Locality specifiers for `do concurrent`

> Locality specified will enable the user to control the data environment of the loop nest in a more fine-grained way. Implementing these specified on the `FIR` dialect level is needed in order to support this in the `DoConcurrentConversionPass`.

Reviewer: ```suggestion
Locality specifiers will enable the user to control the data environment of the loop nest in a more fine-grained way. Implementing these specifiers on the `FIR` dialect level is needed in order to support this in the `DoConcurrentConversionPass`.
```
I see more instances of "specified" that seem to me like they should be "specifier" below as well. Am I reading this wrong perhaps?

Such specified will also unlock a potential solution to the non-perfectly-nested loops' IVs issue described above. In particular, for a non-perfectly nested loop, one middle-ground proposal/solution would be to:
* Emit the loop's IV as shared/mapped just like we do currently.
* Emit a warning that the IV of the loop is emitted as shared/mapped.

> * Given support for `LOCAL`, we can recommend the user to explicitly localize/privatize the loop's  IV if they choose to.

Reviewer: ```suggestion
* Given support for `LOCAL`, we can recommend the user to explicitly localize/privatize the loop's IV if they choose to.
```

### More advanced detection of loop nests

> As pointed out earlier, any intervening code between the headers of 2 nested `do concurrent` loops prevents us currently from detecting this as a loop nest. In some cases this is overly conservative. Therefore, a more flexible detection logic of loop nests needs to be implemented.

Reviewer: Do you have already any plan in mind to decide where to place the operations that were originally located between the outer and inner loop?

Author: Nothing concrete yet. Stay tuned 😛.
